src/riak_kv_qry_buffers.erl

```diff
@@ -420,55 +427,97 @@ do_batch_put(QBufRef, Data, #state{qbufs          = QBufs0,
+            IsReady = (ChunksNeed == ChunksGot),
+
+            %% construct keys for new data
             OrdByFieldQualifiers = lists:map(fun get_ordby_field_qualifiers/1, OrderBy),
-            case riak_kv_qry_buffers_ldb:add_rows(LdbRef, Data, ChunkId,
-                                                  KeyFieldPositions,
-                                                  OrdByFieldQualifiers) of
-                ok ->
-                    IsReady = (ChunksNeed == ChunksGot),
-                    QBuf = QBuf0#qbuf{size          = Size + ChunkSize,
-                                      chunks_got    = ChunksGot,
-                                      total_records = TotalRecords0 + length(Data),
-                                      is_ready      = IsReady,
-                                      last_accessed = os:timestamp()},
+            KeyedData = enkey(Data, KeyFieldPositions, OrdByFieldQualifiers, ChunkId),
+
+            QBuf1 = QBuf0#qbuf{size          = Size + ChunkSize,
+                               chunks_got    = ChunksGot,
+                               total_records = TotalRecords0 + length(Data),
+                               is_ready      = IsReady,
+                               last_accessed = os:timestamp()},
+
+            HaveLdbOpened = not is_function(LdbRefOrDelayedCreateFun),
+            InmemAcc = lists:append(InmemBuffer0, KeyedData),
+            EleveldbPutF =
+                fun(LdbRef) ->
+                        lager:debug("adding chunk ~b of ~b to ldb", [ChunksGot, ChunksNeed]),
+                        case riak_kv_qry_buffers_ldb:add_rows(LdbRef, InmemAcc) of
+                            ok ->
+                                QBuf = QBuf1#qbuf{inmem_buffer = [],
+                                                  ldb_ref      = LdbRef},
+                                {ok, QBuf};
+                            {error, _} = ErrorReason ->
+                                ErrorReason
+                        end
+                end,
+
+            case (HaveLdbOpened orelse not can_afford_inmem(InmemMax)) of
+                false ->
+                    InmemBuffer =
+                        lists:sort(InmemAcc),
+                    lager:debug("adding chunk ~b of ~b to inmem buffer", [ChunksGot, ChunksNeed]),
+                    QBuf = QBuf1#qbuf{inmem_buffer = maybe_only_rows(InmemBuffer, IsReady)},
                     {ok, QBuf};
-                {error, _} = ErrorReason ->
-                    ErrorReason
+                true when HaveLdbOpened ->
+                    EleveldbPutF(LdbRefOrDelayedCreateFun);
+                true ->
+                    case LdbRefOrDelayedCreateFun() of
+                        {ok, LdbRef} ->  %% now it's a real ref
+                            EleveldbPutF(LdbRef);
+                        {error, _} = ErrorReason ->
+                            ErrorReason
+                    end
             end
     end.
 
+can_afford_inmem(Threshold) ->
```

Comment:
Perhaps I'm overlooking the obvious, but why are we not tracking the data size as chunks arrive and comparing against that, instead of incurring the overhead of checking the process information?